src/transformers/models/mask2former/modeling_mask2former.py

```diff
@@ -1302,16 +1312,18 @@ def forward(
         ]
 
         # Prepare encoder inputs (by flattening)
-        spatial_shapes = [(embed.shape[2], embed.shape[3]) for embed in input_embeds]
+        spatial_shapes_list = [(embed.shape[2], embed.shape[3]) for embed in input_embeds]
         input_embeds_flat = torch.cat([embed.flatten(2).transpose(1, 2) for embed in input_embeds], 1)
-        spatial_shapes = torch.as_tensor(spatial_shapes, dtype=torch.long, device=input_embeds_flat.device)
+        spatial_shapes = torch.as_tensor(spatial_shapes_list, dtype=torch.long, device=input_embeds_flat.device)
         masks_flat = torch.cat([mask.flatten(1) for mask in masks], 1)
 
         position_embeddings = [embed.flatten(2).transpose(1, 2) for embed in position_embeddings]
         level_pos_embed_flat = [x + self.level_embed[i].view(1, 1, -1) for i, x in enumerate(position_embeddings)]
         level_pos_embed_flat = torch.cat(level_pos_embed_flat, 1)
 
-        level_start_index = torch.cat((spatial_shapes.new_zeros((1,)), spatial_shapes.prod(1).cumsum(0)[:-1]))
+        level_start_index = [0]
+        for h, w in spatial_shapes_list[:-1]:
+            level_start_index.append(level_start_index[-1] + h * w)
```

Comment:
Is this change necessary for torch compile/export? Does it impact performance? Just curious as this is also present in other models where a spatial_shapes_list is added, but it did not seem to break the graph or cause any problems with torch compile

Reply:
I'll investigate some more, but when I undo this line change, I start getting this error during torch.export
```
from user code:
   File "/home/philkuz/dev/transformers/src/transformers/models/mask2former/modeling_mask2former.py", line 2512, in forward
    outputs = self.model(
  File "/home/philkuz/.pyenv/versions/gml311/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/philkuz/dev/transformers/src/transformers/models/mask2former/modeling_mask2former.py", line 2283, in forward
    pixel_level_module_output = self.pixel_level_module(
  File "/home/philkuz/.pyenv/versions/gml311/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/philkuz/dev/transformers/src/transformers/models/mask2former/modeling_mask2former.py", line 1406, in forward
    decoder_output = self.decoder(backbone_features, output_hidden_states=output_hidden_states)
  File "/home/philkuz/.pyenv/versions/gml311/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/philkuz/dev/transformers/src/transformers/models/mask2former/modeling_mask2former.py", line 1355, in forward
    outputs = [
  File "/home/philkuz/dev/transformers/src/transformers/models/mask2former/modeling_mask2former.py", line 1356, in <listcomp>
    x.transpose(1, 2).view(batch_size, -1, spatial_shapes_list[i][0], spatial_shapes_list[i][1])
 ```
 
 which is connected through this:
 1. `level_start_index` fills in the `split_sizes` in L1347-1352
 2. `split_sizes` is an input to `torch.split` which outputs `encoder_output` in L1345 
 3. `encoder_output` is iterated over in the list comprehension of L1357-1359 which is where we get the bottom of the stack trace.
 
 Let me find the other uses of `spatial_shapes_list` and see what's different.

Reply:
Tangent: `level_start_index` isn't used at all in the attention module and I can't find any other uses than to inform `split_sizes`

Reply:
Some updates:
I've found the following models use a `level_start_index`:
1. deformable_detr
2. oneformer
3. rt_detr
this is not a full list, but I think helped me get an idea of what's going on.

neither deformable_detr nor rt_detr have a similar `torch.split` call which is a way to break apart the `multi-level` result of the decoder
<img width="412" alt="image" src="https://github.com/user-attachments/assets/99bfc2b6-06fd-4993-9c6f-f7e8e0292ef9">

`oneformer` has a simliar setup, but I also failed to export the model, likely because the issues are very similar to what I faced with mask2former. Happy to help make `oneformer` compatible for export if you would like!

But my conclusion is that `level_start_index` does need to change to be compatible for our specific use-cases. We could also simply rename `level_start_index` in this case to better match that we are using it for split rather than for `level_start_index` parameter, especially since we don't actually use this value in the modeling code.

Reply:
Thanks for diving deeply in this issue and for the detailed explanation! You're right it looks like `level_start_index` is actually not used in this model, apart for building `split_sizes`, so we definitely shouldn't let it block `torch.export` :). Looking at other models using `level_start_index`, it seems that it is used only for the custom kernel `MultiScaleDeformableAttentionFunction`, which is not present in mask2former (don't know why, but that can be the object of another PR).
I agree with what you've done, I think we can leave the tensor version of `level_start_index` even if it's not used now, and open a PR in the future to add the custom kernel `MultiScaleDeformableAttentionFunction` that will use it.
As you explained, only oneformer should have a similar problem with `level_start_index`, so feel free to open a PR for that too!